pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -260,13 +263,28 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (result re
 		}
 	}
 
+	info := &reconcileInfo{
+		host:           host,
+		request:        request,
+		bmcCredsSecret: bmcCredsSecret,
+	}
+
+	prov, err := r.provisionerFactory(host, *bmcCreds, info.publishEvent)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	info.cleanSteps = r.buildAndValidateCleanSteps(request, host, prov)
+
 	// Pick the action to perform
 	var actionName metal3v1alpha1.ProvisioningState
 	switch {
 	case !host.Status.GoodCredentials.Match(*bmcCredsSecret):
 		actionName = metal3v1alpha1.StateRegistering
 	case host.WasExternallyProvisioned():
 		actionName = metal3v1alpha1.StateExternallyProvisioned
+	case host.NeedsManualCleaning(info.cleanSteps):
+		actionName = metal3v1alpha1.StateCleaning
```

Comment:
There are manual cleaning and automated cleaning (yes, we're bad at naming things). Manual cleaning is a set of "ready-state" steps, and it should run before inspection, so that inspection sees RAID. Automated cleaning is what wipes the disks, and it runs before the first and after every deployment. I hope that helps.

Reply:
> Should cleaning happen before inspection?

In the case of `cleaning` includes RAID configuration, some Storage details in HardwareDetails might change, so we need `inspecting` phase to refresh these information. 

> I would have expected to do it as part of provisioning and deprovisioning.

IMO, we should not put `cleaning` into `provisioning` phase. I think both of RAID and BIOS should be configured only one time, so if we run `cleaning` before `inspecting` phase, the latest hardware information would be updated in `HardwareDetails`.
But I am confusing whether `cleaning` should be put in `registering` or `inspecting`...

@dtantsur: Thank you for the explanation.

Reply:
The HardwareDetails section of the host status is meant to provide information about physical hardware resources. It does not need to reflect the RAID configuration. If ironic is going to fill in RAID info, we will need to strip that part of the data out, but it would be better if we prevent it from showing up there at all.